mtags/src/main/scala/scala/meta/internal/pc/TypeDefinitionProvider.scala

```diff
@@ -0,0 +1,147 @@
+            .mkString
+            .tokenize
+            .get
+            .tokens
+          val curToken = tokens
+            .dropWhile(_.pos.end < params.offset() - vd.pos.start)
+            .head
+
+          if (ignoredTags.contains(curToken.text))
+            None
+          else
+            Some(vd)
+        case t: Tree => Some(t)
+      }
+      .flatMap {
+        case sel: Select
+            if sel.symbol.asMethod.returnType.typeSymbol.isTypeParameter =>
+          Some(sel.tpe.typeSymbol)
+        case app @ Apply(fun, args) if args.nonEmpty =>
+          val (rStart, rEnd) = (fun.pos.start, args.last.pos.end)
+
+          val txt = app.pos.source.content
+            .slice(rStart, rEnd)
+            .fold("")(_ + _.toString)
+            .toString
+          val tokens = txt.tokenize.get.tokens.toList
+            .filter(t => {
+              val (pStart, pEnd) =
+                (app.pos.start + t.pos.start, app.pos.start + t.pos.end)
+              pStart <= params.offset && pEnd >= params.offset
+            })
+          tokens.headOption match {
+            case Some(t) =>
+              app.symbol.asMethod.paramLists.flatten
+                .find(_.nameString.trim == t.text)
+                .map(_.tpe.typeSymbol)
+            case _ => None
+          }
+        case tree: Tree if tree.symbol.isMethod =>
+          Some(tree.symbol.asMethod.returnType.typeSymbol)
+        case tree: Tree if tree.symbol.isTypeSymbol =>
+          Some(tree.symbol)
+        case tree: Tree if tree.tpe.isDefined =>
+          Some(tree.tpe.typeSymbol)
+        case tree: Tree if tree.children.nonEmpty =>
+          Some(tree.children.head.tpe.typeSymbol)
+        case vd: compiler.ValDef if vd.rhs.tpe != null =>
+          Some(vd.rhs.tpe.typeSymbol)
+        case tree: Tree =>
+          val expTree = expandRangeToEnclosingApply(tree.pos)
+          if (expTree.tpe != null && expTree.tpe.isDefined)
+            Some(expTree.tpe.typeSymbol)
+          else None
+        case _ => None
+      }
+  }
+
+  def typeDefinition(params: OffsetParams): List[l.Location] =
+    typeSymbol(params)
+      .collect {
+        case sym => getSymbolDefinition(sym)
+      }
+      .toList
+      .flatten
+
+  private def getSymbolDefinition(sym: Symbol): List[l.Location] = {
+    val file = sym.pos.source.file
+    val uri =
+      try {
+        file.toURL.toURI.toString
+      } catch {
+        case _: NullPointerException =>
+          sym.pos.source.path
+      }
+
+    try {
+      if (file != null || compiler.unitOfFile.contains(file)) {
+        val unit = compiler.unitOfFile(file)
+        if (unit != null) {
+          val trees = unit.body
+            .filter(_.symbol == sym)
+            .filter(_.pos != null)
+            .filter(_.isDef)
+          trees
+            .map(t => new l.Location(uri, t.pos.toLSP))
+        } else fallbackToSemanticDB(sym)
+      } else fallbackToSemanticDB(sym)
+    } catch {
```

Comment:
There is already a lot of checks, do we need this?

Reply:
Changed.